daemon/run.go

```diff
@@ -0,0 +1,319 @@
+package daemon
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/gorilla/mux"
+
+	"github.com/akitasoftware/akita-cli/apispec"
+	"github.com/akitasoftware/akita-cli/har_loader"
+	"github.com/akitasoftware/akita-cli/plugin"
+	"github.com/akitasoftware/akita-cli/rest"
+	"github.com/akitasoftware/akita-cli/trace"
+	"github.com/akitasoftware/akita-cli/util"
+	"github.com/akitasoftware/akita-libs/akid"
+	kgxapi "github.com/akitasoftware/akita-libs/api_schema"
+	"github.com/akitasoftware/akita-libs/sampled_err"
+)
+
+const WITNESS_BUFFER_SIZE = 10_000
+
+type Witness = har_loader.CustomHAREntry
+
+type Args struct {
+	// Required args.
+	ClientID akid.ClientID
+	Domain   string
+
+	// Optional args.
+	PortNumber uint16
+
+	Plugins []plugin.AkitaPlugin
+}
+
+// A wrapper around HTTP request handlers for handling HTTP responses generically.
+type httpHandler func(*http.Request) HTTPResponse
+
+var cmdArgs Args
+
+func Run(args Args) error {
+	cmdArgs = args
+	router := mux.NewRouter().StrictSlash(true)
+
+	// Register an endpoint for creating a new learning session.
+	router.Handle("/services/{serviceName}/learn/sessions", httpHandler(createLearnSession)).Methods("POST")
+
+	// Register an endpoint for adding witnesses to a learning session. The request body is expected to be a stream of HAR entry objects to be added.
+	router.Handle("/services/{serviceName}/learn/sessions/{learnSessionName}/witnesses", httpHandler(addWitnesses)).Methods("POST")
+
+	// Register an endpoint for creating an API model out of a set of learning sessions. The request body is expected to be a JSON object specifying the names of the learning sessions.
+	router.Handle("/services/{serviceName}/api-models", httpHandler(createModel)).Methods("POST")
+
+	listenSocket := fmt.Sprintf("127.0.0.1:%d", cmdArgs.PortNumber)
+	log.Fatal(http.ListenAndServe(listenSocket, router))
+	return nil
+}
+
+// Obtains the service ID for the service name contained in the given HTTP request variables. If an error occurs, this is formatted and returned as an HTTP response.
+func getServiceID(requestVars map[string]string) (akid.ServiceID, *HTTPResponse) {
+	var httpErr *HTTPResponse = nil
+
+	serviceName := requestVars["serviceName"]
+	frontClient := rest.NewFrontClient(cmdArgs.Domain, cmdArgs.ClientID)
+	result, err := util.GetServiceIDByName(frontClient, serviceName)
+	if err != nil {
+		httpErr = NewHTTPError(err, http.StatusNotFound, "Service not found")
+	}
+	return result, httpErr
+}
+
+// Creates a new learning session in the Akita back end.
+func createLearnSession(request *http.Request) HTTPResponse {
+	vars := mux.Vars(request)
+
+	// Get the service ID.
+	serviceID, httpErr := getServiceID(vars)
+	if httpErr != nil {
+		return *httpErr
+	}
+
+	// Create a new learning session with a random name.
+	// TODO: Allow user to specify a name?
```

Comment:
Yes, this would be nice to add down the road.  Not necessary for this PR.

Reply:
Done in 2231328.